File: src/main/kotlin/easy/MergeTwoSortedList.kt
package easy

import datastructures.ListNode

class MergeTwoSortedList {
    /*
    * Link: https://leetcode.com/problems/merge-two-sorted-lists/
    * Difficulty: Easy
    * */
    fun mergeTwoLists(list1: ListNode?, list2: ListNode?): ListNode? {
        // Como as duas já estão ordenadas, podemos percorrer o array e
        // comparar cada elemento e decidir qual o maior
        var ll1 = list1
        var ll2 = list2

        // Esse algorítmo tem complexidade 0(N) de tempo no qual N é a quantidade
        // de elementos de cada lista somados enquanto o uso de memória é constante
        val head = ListNode(0)
        var temp = head
        while (ll1 != null || ll2 != null) {
            if((ll1?.`val` ?: 999) <= (ll2?.`val` ?: 999)){
                temp.next = ll1
                ll1 = ll1?.next
            } else {
                temp.next = ll2
                ll2 = ll2?.next
            }
            temp = temp.next!!
        }

        return head.next
    }
}

File: src/easy/merge_two_sorted_list.go
package easy

import "github.com/johnazedo/leetcode/src"

/*
Difficult: Easy
Link: https://leetcode.com/problems/merge-two-sorted-lists
*/
func mergeTwoLists(list1 *src.ListNode, list2 *src.ListNode) *src.ListNode {
	head := &src.ListNode{}
	var v1 int
	var v2 int
	temp := head
	for list1 != nil || list2 != nil {
		if list1 != nil {
			v1 = list1.Val
		} else {
			v1 = 999
		}
		if list2 != nil {
			v2 = list2.Val
		} else {
			v2 = 999
		}

		if v1 <= v2 {
			temp.Next = list1
			list1 = list1.Next
		} else {
			temp.Next = list2
			list2 = list2.Next
		}
		temp = temp.Next

	}
	return head.Next
}